heroku/heroku_2fa.py

```diff
@@ -0,0 +1,172 @@
+session = requests.session()
+
+
+def have_creds():
+    return get_netrc_auth(ORG_USERS_URL)
+
+
+def find_users_missing_2fa():
+    users_missing_2fa = {}
+    try:
+        org_users = fetch_api_json(ORG_USERS_URL)
+        users_missing_2fa = defaultdict(set)
+        for user in org_users:
+            if not user['two_factor_authentication']:
+                users_missing_2fa[user['role']].add(user['email'])
+    except Exception:
+        logger.critical("Failure communicating with Heroku", exc_info=True)
+        update_exit_code(3)
+    return users_missing_2fa
+
+
+def apps_accessible_by_user(email, role):
+    if role == 'admin':
+        return ['ALL']
+    users_apps_url = '{}/{}/apps'.format(ORG_USERS_URL, email)
+    return [app['name'] for app in fetch_api_json(users_apps_url)]
+
+
+def fetch_api_json(url):
+    # The requests library will automatically use credentials found in netrc.
+    response = session.get(url, headers=REQUEST_HEADERS, timeout=30)
+    response.raise_for_status()
+    return response.json()
+
+
+def find_affected_apps(users_missing_2fa):
+    affected_apps = defaultdict(set)
+    for role, users in users_missing_2fa.items():
+        # print('\n~ {} {}s:'.format(len(users), role))
+        for email in sorted(users):
+            for app in apps_accessible_by_user(email, role):
+                affected_apps[app].add(email)
+            # print(email)
+    return affected_apps
+
+
+def generate_csv(users_missing_2fa, affected_apps):
+    if affected_apps:
+        print('"{}","{}"'.format("Application", "User without 2FA"))
+        for app, emails in sorted(affected_apps.items()):
+            for email in emails:
+                print('"{}","{}"'.format(app, email))
+    return
+
+
+def output_results(users_missing_2fa, affected_apps):
+    if not users_missing_2fa:
+        print('All {} users have 2FA enabled :)'.format(ORG_NAME))
+        return
+
+    print('The following {} users do not have 2FA enabled!'.format(ORG_NAME))
+    for role, users in users_missing_2fa.items():
+        print('\n~ {} {}s:'.format(len(users), role))
+        for email in sorted(users):
+            print(email)
+
+    if affected_apps:
+        print('\n{} apps are affected:\n'.format(len(affected_apps)))
+        for app, emails in sorted(affected_apps.items()):
+            print('{} ({})'.format(app, ', '.join(sorted(emails))))
+
+
+def main(args):
+    if not get_netrc_auth(ORG_USERS_URL):
+        print('Heroku API credentials not found in `~/.netrc` or `~/_netrc`.\n'
+              'Log in using the Heroku CLI to generate them.')
+        update_exit_code(1)
+        return
+
+    users_missing_2fa = find_users_missing_2fa()
+    affected_apps = find_affected_apps(users_missing_2fa)
+
+    if args.csv:
+        generate_csv(users_missing_2fa, affected_apps)
+    else:
+        output_results(users_missing_2fa, affected_apps)
+
+    update_exit_code(0 if not users_missing_2fa else 2)
+
+
+def parse_args():
```

Comment:
as above, this Python module can be used from the cli or imported. the cli support functions are not used. Yes, a full python package would be better -- happy to take that as a Phase 2, rather than blocker.

also which ``bin/``? ``heroku/bin``? Is that automatically on the python path so I can import?

Reply:
How about:

* move this script out to avoid duplicating it
* run it beforehand to create the CSV
* pass the CSV path as a config param and skip this test when the param isn't present
* read the CSV in the test or `resources.py`

?